src/main/kotlin/com/serameracorp/pattern/PatternService.kt

```diff
@@ -0,0 +1,140 @@
+
+class PatternService {
+
+    val patternRepository = PatternRepository()
+    fun createPattern(formParams: Parameters): Int {
+        val nameParam = formParams["name"]
+        val publisherParam = formParams["publisher"]
+        val publishedInParam = formParams["published_in"]
+        val difficultyParam = formParams["difficulty"]
+
+        patternRepository.createPatternStatement.setString(1, nameParam ?: "")
+        patternRepository.createPatternStatement.setString(2, publisherParam ?: "")
+        patternRepository.createPatternStatement.setString(3, difficultyParam ?: "")
+        patternRepository.createPatternStatement.setString(4, publishedInParam ?: "")
+        val resultSet = patternRepository.createPatternStatement.executeQuery()
+        resultSet.next()
+        return resultSet.getInt("id")
+    }
+
+    fun createPatternFabrics(patternId: Int, formParams: Parameters) {
+        for (x in 0..2) {
+            val fabricParam = formParams["fabric$x"]
+            val fabricLengthParam = formParams["fabric${x}_length"]
+            createPatternFabric(patternId, fabricParam, fabricLengthParam)
+        }
+    }
+
+    fun createPatternFabric(patternId: Int, fabricParam: String?, fabricLengthParam: String?) {
+        if (fabricParam != null && fabricParam != "") {
+            val fabricId = fabricParam.toIntOrNull()
+            if (fabricId == null) {
+                throw IllegalArgumentException("Invalid fabricId: '$fabricParam'")
+            }
+            val fabricLength = fabricLengthParam?.toDoubleOrNull()
+            if (fabricLength == null) {
+                throw IllegalArgumentException("Invalid fabricLength: '$fabricLengthParam'")
+            }
+            patternRepository.createPatternFabricStatement.setInt(1, patternId)
+            patternRepository.createPatternFabricStatement.setInt(2, fabricId)
+            patternRepository.createPatternFabricStatement.setDouble(3, fabricLength)
+            patternRepository.createPatternFabricStatement.executeQuery()
+        }
+    }
+
+    fun findAllPatternsByName(searchParam: String?): List<Pattern> {
+        val resultSet = patternRepository.searchPatternStatement(searchParam).executeQuery()
+
+        return sequence {
+            while (resultSet.next()) {
+                yield(patternFromResultSet(resultSet))
+            }
+        }.toList()
+    }
+
+    fun findPatternWithDetails(patternId: Int): Pattern{
+        val pattern = findPatternById(patternId)
+        addPatternDetails(pattern)
+        return pattern
+    }
+
+    fun findPatternById(patternId: Int): Pattern {
+        patternRepository.patternByIdStatement.setInt(1, patternId)
+        val resultSet = patternRepository.patternByIdStatement.executeQuery()
+        if (resultSet.next()) {
+            // no fabric details yet
+            return patternDetailsFromResultSet(resultSet)
+        } else {
+            throw NotFoundException("The searched pattern was not found")
```

Comment:
I'd prefer returning a `Pattern?` here instead of throwing an exception. It's not 'exceptional' behaviour for an unknown pattern id to appear.